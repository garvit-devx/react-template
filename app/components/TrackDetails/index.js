/**
 *
 * TrackDetails
 *
 */

import React, { useRef, useState } from 'react';
import PropTypes from 'prop-types';
import Button from '@mui/material/Button';
import styled from '@emotion/styled';
import T from '@components/T';

const TrackDetailsContainer = styled.div`
  display: flex;
  gap: 2rem;
  max-width: 720px;
  margin: 3em auto;
`;

const TrackImage = styled.img`
  width: 100%;
  height: auto;
  object-fit: cover;
  aspect-ratio: 1/1;
`;

export function TrackDetails({ trackDetails }) {
  const audioRef = useRef(null);
  const [isPlaying, setIsPlaying] = useState(false);
  const { artistName, artworkUrl100, previewUrl, trackName, trackTimeMillis } = trackDetails;
  const trackTimeInMinutes = (trackTimeMillis / 1000 / 60).toFixed(2);

  function handleOnClick() {
    if (isPlaying) {
      audioRef.current.pause();
    } else {
      audioRef.current.play();
    }

    setIsPlaying((currentValue) => !currentValue);
  }

  return (
    <TrackDetailsContainer>
      <div style={{ width: '300px' }}>
        <TrackImage src={artworkUrl100} alt={trackName} />
      </div>

      <div>
        <T id="track_name" values={{ trackName }} />
        <T id="track_artist" values={{ artistName }} />
        <T id="track_time" values={{ trackTimeInMinutes }} />

        <Button variant="secondary" sx={{ border: '1px solid black', margin: '1rem 0' }} onClick={handleOnClick}>
          {isPlaying ? <T id="pause" /> : <T id="play_preview" />}
        </Button>
        <audio src={previewUrl} ref={audioRef} onEnded={() => setIsPlaying(false)} />
      </div>
    </TrackDetailsContainer>
  );
}

TrackDetails.propTypes = {
  trackDetails: PropTypes.shape({
    artistName: PropTypes.string,
    artworkUrl100: PropTypes.string,
    previewUrl: PropTypes.string,
    trackName: PropTypes.string,
    trackTimeMillis: PropTypes.number
  })
};

export default TrackDetails;
