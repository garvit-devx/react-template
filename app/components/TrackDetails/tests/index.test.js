/**
 *
 * Tests for TrackDetails
 *
 */

import React from 'react';
import { fireEvent } from '@testing-library/dom';
import { renderWithIntl, timeout } from '@utils/testUtils';
import TrackDetails from '../index';

const trackDetails = {
  trackName: 'Perfect',
  artistName: 'One Direction',
  previewUrl: 'Some URL',
  artworkUrl100: 'Some URL',
  trackTimeMillis: 123456789
};

describe('<TrackDetails />', () => {
  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithIntl(<TrackDetails trackDetails={trackDetails} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should contain 1 TrackDetails component', () => {
    const { getAllByTestId } = renderWithIntl(<TrackDetails trackDetails={trackDetails} />);
    expect(getAllByTestId('track-details').length).toBe(1);
  });

  it('should render track details component with all necessary elements', () => {
    const { getByTestId, getByRole, getByText } = renderWithIntl(<TrackDetails trackDetails={trackDetails} />);

    const playButton = getByRole('button', {
      value: { text: 'Play Preview' }
    });
    const audioElement = getByTestId('audio');
    const tackImage = getByRole('img');
    const trackName = getByText(/perfect/i);
    const artistName = getByText(/one direction/i);

    expect(playButton).toBeInTheDocument();
    expect(audioElement).toBeInTheDocument();
    expect(tackImage).toBeInTheDocument();
    expect(trackName).toBeInTheDocument();
    expect(artistName).toBeInTheDocument();
  });

  it('should change button text if button is clicked', async () => {
    const { getByRole } = renderWithIntl(<TrackDetails trackDetails={trackDetails} />);

    let button = getByRole('button', {
      value: { text: 'Play Preview' }
    });
    expect(button).toBeInTheDocument();

    // Text should change to pause if button is clicked
    fireEvent.click(button);
    await timeout(100);
    button = getByRole('button', {
      value: { text: 'Pause' }
    });
    expect(button).toBeInTheDocument();

    // text should change back to 'Play preview' if button is clicked again
    fireEvent.click(button);
    await timeout(100);
    button = getByRole('button', {
      value: { text: 'Play Preview' }
    });
    expect(button).toBeInTheDocument();
  });

  it('should change button text to Play Preview when audio preview ends', async () => {
    const { getByRole, getByTestId } = renderWithIntl(<TrackDetails trackDetails={trackDetails} />);

    let button;

    const audioElement = getByTestId('audio');

    fireEvent(audioElement, new Event('play'));
    await timeout(1000);

    button = getByRole('button', {
      value: { text: 'Pause' }
    });
    expect(button).toBeInTheDocument();

    fireEvent(audioElement, new Event('ended'));
    await timeout(100);

    button = getByRole('button', {
      value: { text: 'Play Preview' }
    });
    expect(button).toBeInTheDocument();
  });
});
