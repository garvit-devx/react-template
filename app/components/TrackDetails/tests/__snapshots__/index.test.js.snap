// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackDetails /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="css-rr8yu3 eh9f6zb1"
      data-testid="track-details"
    >
      <div
        style="width: 300px;"
      >
        <img
          alt="Perfect"
          class="css-1bfxc3 eh9f6zb0"
          src="Some URL"
        />
      </div>
      <div>
        <p
          class="css-lh9t18 egtqi0h0"
          data-testid="t"
        >
          Track Name: Perfect
        </p>
        <p
          class="css-lh9t18 egtqi0h0"
          data-testid="t"
        >
          Artist: One Direction
        </p>
        <p
          class="css-lh9t18 egtqi0h0"
          data-testid="t"
        >
          Track duration: 2057.61 minutes
        </p>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-secondary MuiButton-secondaryPrimary MuiButton-sizeMedium MuiButton-secondarySizeMedium MuiButton-root MuiButton-secondary MuiButton-secondaryPrimary MuiButton-sizeMedium MuiButton-secondarySizeMedium css-3uxdid-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <p
            class="css-lh9t18 egtqi0h0"
            data-testid="t"
          >
            Play Preview
          </p>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <audio
          data-testid="audio"
          src="Some URL"
        />
      </div>
    </div>
  </div>
</body>
`;
